File: algorithms/sort/heap.cpp
#include<bits/stdc++.h>
using namespace std;

void min_heapify(vector<int> &arr, int i){
  int n = arr.size();
  int left = 2*i;
  int right = left + 1;
  int smallest = i;
  if(left < n && arr[i] > arr[left]) smallest = left;
  if(right < n && arr[smallest] > arr[right]) smallest = right;

  if(smallest != i){
    swap(arr[i], arr[smallest]);
    min_heapify(arr, smallest);
  }
}

void build_min_heap(vector<int> &arr){
  int n = arr.size();
  for(int i = n/2; i >= 0; i--){
    min_heapify(arr, i);
  }
}

void heap_sort(vector<int> &arr){
  build_min_heap(arr);
  for (int i = 1; i < arr.size(); i++){
    min_heapify(arr, i);
  }
  
}

int main(){
  vector<int> arr = {8,4,3,10,1,9,7};
  heap_sort(arr);
  for(int i: arr) cout << i << " ";
}

File: algorithms/sort/heap.py
def max_heapify(list, i, n):
  left = 2*i
  right = left + 1
  largest = i

  if left < n and list[i] < list[left]: largest = left
  if right < n and list[largest] < list[right]: largest = right

  if largest != i:
    list[largest], list[i] = list[i], list[largest]
    max_heapify(list, largest, n)

def build_max_heap(list):
  for i in range(len(list)//2, -1, -1):
    max_heapify(list, i, len(list))

def heap_sort(list):
  build_max_heap(list)
  for i in range(len(list)-1, 0, -1):
    list[i], list[0] = list[0], list[i]
    max_heapify(list, 0, i)

list = [4, 5, 9, 8, 7, 3, 2, 1, 6]
heap_sort(list)
print(list)
